RFD 99: Making bundled tsh available outside of Connect

ravicious opened the pull request:
I figured I might as well write down and document what I found before I actually sit down to implement this.

[Rendered version](https://github.com/gravitational/teleport/blob/rfd/0099-link-bundled-tsh/rfd/0099-link-bundled-tsh.md).

Implemented in gravitational/webapps#1445.

ravicious committed:
Add RFD 99: Making bundled tsh available outside of Connect

zmb3 commented on rfd/0099-link-bundled-tsh.md:
```diff
@@ -0,0 +1,127 @@
+
+Instead of symlinking the binary to `/usr/bin/tsh`, we can symlink it to `/usr/local/bin/tsh`. This
+is where our Linux teleport package places our binaries. This way if someone installs Connect after
+installing teleport, we'll overwrite just a single symlink instead of the user ending up with two
```
Might be worth giving the user a message or some other feedback when this is the case.

avatus replied:
> This way if someone installs Connect after installing teleport

What happens if someone updates teleport after they've installed Connect? Could this mess with the protobuf files if the updated tsh contains changes that the Connect binary hasn't shipped with? We haven't had to worry about this with Connect yet since we always ship it with it's "own" tsh, but will this change cause a way to get out of sync?

zmb3 commented on rfd/0099-link-bundled-tsh.md:
```diff
@@ -0,0 +1,127 @@
+plugin](https://nsis.sourceforge.io/EnVar_plug-in) for path manipulation instead due to length
+limitations in the built-in utilities.
+
+We will have to figure out how to use that plugin but once that's done, we should be able to simply
```
Just so I understand correctly, in Windows we're planning on updating the path to point at the bundled tsh rather than creating a binary or symlink somewhere that's already in the path, right?

ravicious replied:
> Just so I understand correctly, in Windows we're planning on updating the path to point at the bundled tsh rather than creating a binary or symlink somewhere that's already in the path, right?

Yes, exactly!

zmb3 commented on rfd/0099-link-bundled-tsh.md:
```diff
@@ -0,0 +1,127 @@
+
+#### macOS
+
+For now we've decided to introduce a custom `tsh install` command to the suggested commands in the
```
This is a little confusing since it will look like a real tsh command, but it's fine for now.

What would we need to do in order to make the command bar more like a VS Code style command palette rather than just a tsh command executor?

ravicious replied:
> What would we need to do in order to make the command bar more like a VS Code style command palette rather than just a tsh command executor?

The first version used to work like the command palette, the first think I worked on was ripping it out and replacing it with autocomplete hehe.

The command palette version was fundamentally incompatible with autocomplete AFAIK. But we could do the same thing VSCode does: make it perform autocomplete by default but if you start the input with `>` then it becomes we replace it with the initial implementation and it becomes a command palette. I think we even considered that at one point.

avatus commented on rfd/0099-link-bundled-tsh.md:
```diff
@@ -0,0 +1,127 @@
+
+#### macOS
+
+For now we've decided to introduce a custom `tsh install` command to the suggested commands in the
```
Will this command be visible after the user has run the command (or has an operating system that runs it automatically)?

ravicious replied:
> Will this command be visible after the user has run the command (or has an operating system that runs it automatically)?

VSCode on macOS always shows both commands. On systems that perform this action during installation, the command is not present. I think we should do the same.

I forgot to mention that the first version of the command bar, back when it worked like a command palette, used to have short descriptions for the commands. We could bring that back to explain what `tsh install`/`tsh uninstall` does.

gzdunek reviewed (approved):
Are we going to inform the user somehow that we updated the PATH? (Maybe by showing a notification)

ravicious commented:
> Are we going to inform the user somehow that we updated the PATH? (Maybe by showing a notification)

Do you mean updating the path on Windows or do you mean informing the user about tsh availability on systems where it's made available automatically during the installation?

I don't think it's necessary to do so on Windows. We distribute tsh as a standalone executable so if we append the bin folder to Path rather than prepending it we should be fine: if the user made a conscious decision to download standalone tsh.exe and put it in the path, we'll respect that. If they don't then tsh will be available straight away.

On Linux, based on Zac's and Michael's comments I think we could echo some messages but only in edge cases, not in the default happy path (where tsh is not already symlinked to `/usr/local/bin/tsh`).

ravicious committed:
Add considerations around existing tsh symlink

ravicious committed:
Fix tables at the top of Connect RFDs

ravicious committed:
Remove warning about overwriting the symlink on macOS

Unlike symlinking during installation, running `tsh install` is a conscious
action that the user makes. I feel that in this case it's not necessary
to emit that warning.

ravicious committed:
Change state to implemented

ravicious committed:
Merge branch 'master' into rfd/0099-link-bundled-tsh
